File: src/main/java/net/sf/rails/game/state/Manager.java
package net.sf.rails.game.state;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * A Manager is a baseline implementation of a Context
 */
public abstract class Manager extends Context {

    // item fields
    private final String id;
    private final Item parent;
    // context fields
    private final Root root;
    private final String fullURI;

    protected Manager(Item parent, String id) {
        checkNotNull(id, "Id cannot be null");
        this.id = id;

        // check arguments, parent can only be null for Root
        checkNotNull(parent, "Parent cannot be null");
        this.parent = parent;

        // URI defined recursively
        fullURI =  parent.getFullURI() + Item.SEP + id;

        // find root and add context there
        root = parent.getContext().getRoot();
        // add to root
        root.addItem(this);
    }

    @Override
    public String getId() {
        return id;
    }

    @Override
    public Item getParent() {
        return parent;
    }

    @Override
    public Context getContext() {
        if (parent instanceof Context) {
            return (Context)parent;
        } else {
            // recursive definition
            return parent.getContext();
        }
    }

    @Override
    public String getURI() {
        if (parent instanceof Context) {
            return id;
        } else {
            // recursive definition
            return parent.getURI() + Item.SEP + id;
        }
    }

    @Override
    public String getFullURI() {
        return fullURI;
    }

    @Override
    public String toText() {
        return id;
    }

    // Context methods
    @Override
    public Item locate(String uri) {
        // first try as fullURI
        Item item = root.locateFullURI(uri);
        if (item != null) return item;
        // otherwise as local
        return root.locateFullURI(fullURI + Item.SEP + uri);
    }


   @Override
   void addItem(Item item) {
        // check if this context is the containing one
        checkArgument(item.getContext() == this, "Context is not the container of the item to add");

        // add item to root
        root.addItem(item);
   }

   @Override
   void removeItem(Item item) {
       // check if this context is the containing one
       checkArgument(item.getContext() == this, "Context is not the container of the item to add");

       // remove item from root
       root.removeItem(item);
   }

    @Override
    public Root getRoot() {
        return root;
    }
}

File: dotnet/RailsLib.Net/Net/Game/State/Manager.cs
using GameLib.Net.Util;
using System;

namespace GameLib.Net.Game.State
{
    public class Manager : Context
    {
        // item fields
        private readonly string id;
    private readonly IItem parent;
    // context fields
    private readonly Root root;
    private readonly string fullURI;

    protected Manager(IItem parent, string id)
        {
            Precondition.CheckNotNull(id, "Id cannot be null");
            this.id = id;

            // check arguments, parent can only be null for Root
            Precondition.CheckNotNull(parent, "Parent cannot be null");
            this.parent = parent;

            // URI defined recursively
            fullURI = parent.FullURI + IItemConsts.SEP + id;

            // find root and add context there
            root = parent.GetContext.GetRoot;
            // add to root
            root.AddItem(this);
        }

        override public string Id
        {
            get
            {
                return id;
            }
        }

        override public IItem Parent
        {
            get
            {
                return parent;
            }
        }

        override public Context GetContext
        {
            get
            {
                if (parent is Context)
                {
                    return (Context)parent;
                }
                else
                {
                    // recursive definition
                    return parent.GetContext;
                }
            }
        }

        override public string URI
        {
            get
            {
                if (parent is Context)
                {
                    return id;
                }
                else
                {
                    // recursive definition
                    return parent.URI + IItemConsts.SEP + id;
                }
            }
        }

        override public string FullURI
        {
            get
            {
                return fullURI;
            }
        }

        override public string ToText()
        {
            return id;
        }

        // Context methods
        override public IItem Locate(string uri)
        {
            // first try as fullURI
            IItem item = root.LocateFullURI(uri);
            if (item != null) return item;
            // otherwise as local
            return root.LocateFullURI(fullURI + IItemConsts.SEP + uri);
        }

        override public void AddItem(IItem item)
        {
            // check if this context is the containing one
            Precondition.CheckArgument(item.GetContext == this, "Context is not the container of the item to add");

            // add item to root
            root.AddItem(item);
        }

        override public void RemoveItem(IItem item)
        {
            // check if this context is the containing one
            Precondition.CheckArgument(item.GetContext == this, "Context is not the container of the item to add");

            // remove item from root
            root.RemoveItem(item);
        }

        override public Root GetRoot
        {
            get
            {
                return root;
            }
        }
    }
}
